Source/Core/Event.js

```diff
@@ -7,6 +7,7 @@ import defined from "./defined.js";
  * exposed as a property for others to subscribe to.
  *
  * @alias Event
+ * @template Listener extends Function = Function
```

Comment:
I can't fine `@template` in the official JSDoc documentation, https://jsdoc.app/.

It this not actually valid JSDoc and instead a Closure compiler extension? (as mentioned here https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#template-t)

Also, the `extends Function = Function` thing working almost looks like a happy accident.

I'm still learning my way around TS generics, but does this basically let non-templated versions of Even still work with a default Function callback that takes any number/type of arguments?

Reply:
Ideally the Event class's definition would look like this:

```typescript
export class Event<Listener extends (...args: any[]) => any = (...args: any[]) => any> {
    constructor();
    readonly numberOfListeners: number;
    addEventListener(listener: Listener, scope?: any): Event.RemoveCallback;
    removeEventListener(listener: Listener, scope?: any): boolean;
    raiseEvent(...arguments: Parameters<Listener>): void;
}
```

Using `(...args: any[]) => any` is better than `Function` because it actually has a call signature, unlike `Function`. So you can't use `Function` with `Parameters<Listener>` to get the nicely-typed signature for that function.

I just pushed up a change to this, and it seems to work ok. Even the generated reference doc looks pretty good, even if we're abusing it a bit. But let me know if I've missed something.

Reply:
You're right, `@template` is a Closure extension -- look at the very bottom of the `tsd-jsdoc` README.  That's also the bit where it said you need the plugin to make it work.  There are also a bunch of different competing syntaxes for function signatures: `Function.<number, boolean>`, `function(number):boolean`, `(n: number) => boolean`.  I'm not sure which of those is "normal" JSDoc, which is comes from Closure, etc.  I know `Function` (without constraints) is fine for everybody.

I believe in my experience that the TS compiler treats `Function` and `(...args: any[]) => any` as equivalent, sort of like `any` vs `{[key: any]: any}`, though there may be some nuance I'm not familiar with.  A more specific call signature, like `(n: number, b: boolean) => void` should be assignable to either.  Also, these are event listeners, so I believe the return type should always be `void` -- there's no way to do anything with the function result, right?

I actually originally wanted to make Event generic on `<...args: any[]>`, which would make typing `raiseEvent` easier, but [generic rest parameters](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#generic-rest-parameters) are a relatively recent Typescript addition and I can't even begin to imagine how you'd document them as valid JSDoc.  Full disclosure: I wasn't too worried about `raiseEvent` because I didn't expect users to be raising events, just consuming them.

Reply:
Function isn't quite the same in that you can't use it with `Parameters<T>`. This TypeScript playground shows what I mean:

https://www.typescriptlang.org/play/index.html#code/MYGwhgzhAECiBuBTAdgFwPICcCWBzbyYIAPADLYSoqKbSIAeVyAJjAGICuywq2A9smgBeaJ268BAPmgBvAFDRF0MM2YIUqcpWqYAFCApMaALmhajmADTQIwPgAdEAflNhkATwCUp+H2zNZOQUlAF9gxUwwCkR1NF0AOkSwTFwIUwAFZLAAW0QqTAgyQx1Jb2hff0Dw6DCwuVBIGFjUAFk+ZmwAM2xEZiLtZBo6RhRWaASklLTlDwBtAF1PYWk3d2FxxPjk1Nc5xeWZ92l5JWVVZvMdfWLBzFNL2+tbB2ddrx8-APlqsNPI6OaEy2UwyWVy+UKDxopQ+lW+p1qQVQ7kc0Ba7jEwHWunopmQHGyACMaNZ3KZKDhkLglkJpPiiTQANxBOzISh0daDADucCQaDaHW6vWI6Mxkl0nmZiC25z5mhuNF0OLxBOJVmgZJsqEp1IO9ElcmlKjUcqhemV0Hpapp0n1Uvi-wgMTlugALNYAEQAZg9BqNso0ZqVuK1Opt0DthodUSdgI9VEoHusAEYDUA

Also note that nice static typing on `raiseEvent` with my new version, even though you're right that most CesiumJS users won't be raising events. The last two lines in the TS playground are intentionally errors to show that the calls are type checked correctly.

It all seems to work ok with tsd-jsdoc and with jsdoc itself, unless you spot a problem I didn't.

Reply:
> Also, these are event listeners, so I believe the return type should always be void

Yeah that's a good point. Certainly raiseEvent ignores the return values and returns void itself. No major harm in the listeners returning non-void, but it's a little misleading.

Reply:
> there may be some nuance I'm not familiar with

You found it! 😁 

Seriously though, this is great, as long as we don't wind up with some kind of issue with other consumers of the docs -- like I said, I'm not really clear on who wants `function` vs `Function` vs `Function.<T,U,V>` vs `() => Whatever` and I was worried that the arrow-notation version was "too Typescript-y".  If the team is happy with it, I certainly am too.